hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/metrics/RBFMetrics.java

```diff
@@ -372,12 +420,69 @@ private static void setStateStoreVersions(
 
   @Override
   public long getTotalCapacity() {
-    return getNameserviceAggregatedLong(MembershipStats::getTotalSpace);
+    return getNameserviceAggregatedLong(
+        DatanodeReportType.LIVE, DatanodeInfo::getCapacity);
+  }
+
+  public LoadingCache<DatanodeReportType, DatanodeInfo[]> getDnCache() {
+    return dnCache;
+  }
+
+  /**
+   * Get the aggregated value for a DatanodeReportType and
+   * a method for all nameservices.
+   * @param type a DatanodeReportType
+   * @param f Method reference
+   * @return Aggregated long.
+   */
+  public long getNameserviceAggregatedLong(
+      DatanodeReportType type, ToLongFunction<DatanodeInfo> f){
+    long size = 0;
+    try {
+      size = Arrays.stream(dnCache.get(type)).mapToLong(f).sum();
+    } catch (ExecutionException e) {
+      LOG.debug("Cannot get " + type + " nodes", e.getMessage());
+    }
+    return size;
+  }
+
+  /**
+   * Get the aggregated value for a DatanodeReportType and
+   * a method for all nameservices.
+   * @param type a DatanodeReportType
+   * @param f Method reference
+   * @return Aggregated Integer.
+   */
+  public int getNameserviceAggregatedInt(
+      DatanodeReportType type, Predicate<DatanodeInfo> f){
+    int size = 0;
+    try {
+      Arrays.stream(dnCache.get(DatanodeReportType.LIVE)).filter(f).count();
```

Comment:
We are not updating size at all, are we?
It is also not the most intuitive code to read; maybe extract a litle.